### PR TITLE
[ResultBuilders] Properly diagnose `weak` declarations with non-optional type

### PR DESCRIPTION
**Pull request by xedin**
Diagnose situations where `weak` attribute is used with a non-optional type
in declaration used in the body of a result builder e.g. `weak var x: <Type>`:

```swift
class X {
}

// some result builder context
weak var x: X = ...
```

`weak` declaration is required to use an optional type e.g. `X?`.

### CODE REVIEW

File: test/Constraints/result_builder_diags.swift [hunk 1]
```diff
@@ -781,3 +781,21 @@ func test_rdar65667992() {
...
+
+func test_weak_with_nonoptional_type() {
+  class X {
+    func test() -> Int { 0 }
+  }
+
+  tuplify(true) { c in
+    // FIXME(diagnostic): The warning is incorrect, because `weak` cannot be declared as `let`.
+    weak var x: X = X() // expected-error {{'weak' variable should have optional type 'X?'}}
+    // expected-warning@-1 {{variable 'x' was never mutated; consider changing to 'let' constant}}
```
Comment (xedin) [inline test/Constraints/result_builder_diags.swift:793]: @hborla I'd looking into why structural diagnostics are run for invalid result builder bodies separately.

**Comment by xedin** [comment]
@swift-ci please smoke test

**Comment by xedin** [comment]
@swift-ci please smoke test

**Comment by xedin** [comment]
@swift-ci please smoke test Linux platform

**Review by hborla — APPROVED** [review]

**Pull request merged**